test/PartialRepayments.ts

```diff
@@ -0,0 +1,1605 @@
+            await mockERC20.connect(borrower).approve(loanCore.address, total);
+
+            expect(await vaultFactory.ownerOf(bundleId)).to.eq(loanCore.address);
+
+            // go to 1 block before loan expires
+            await blockchainTime.increaseTime((31536000 / 2) - 3);
+
+            await expect(
+                repaymentController.connect(borrower).repay(loanId, ethers.constants.MaxUint256)
+            ).to.emit(loanCore, "LoanRepaid").withArgs(loanId);
+
+            expect(await mockERC20.balanceOf(borrower.address)).to.eq(0);
+        });
+
+        it("Force repay full. 100 ETH principal, 10% interest rate, full duration.", async () => {
+            const { repaymentController, vaultFactory, mockERC20, loanCore, borrower, blockchainTime } = ctx;
+
+            const { loanId, bundleId } = await initializeLoan(
+                ctx,
+                mockERC20.address,
+                BigNumber.from(31536000), // durationSecs
+                ethers.utils.parseEther("100"), // principal
+                1000, // interest
+                1754884800, // deadline
+            );
+
+            // total repayment amount
+            const total = ethers.utils.parseEther("110");
+            const repayAdditionalAmount = total.sub(await mockERC20.balanceOf(borrower.address));
+            // mint borrower exactly enough to repay loan
+            await mint(mockERC20, borrower, repayAdditionalAmount);
+            await mockERC20.connect(borrower).approve(loanCore.address, total);
+
+            expect(await vaultFactory.ownerOf(bundleId)).to.eq(loanCore.address);
+
+            // go to 1 block before loan expires
+            await blockchainTime.increaseTime(31536000 - 3);
+
+            await expect(
+                repaymentController.connect(borrower).forceRepay(loanId, ethers.constants.MaxUint256)
+            ).to.emit(loanCore, "LoanRepaid").withArgs(loanId);
+
+            expect(await mockERC20.balanceOf(borrower.address)).to.eq(0);
+
+            const noteReceipt = await loanCore.noteReceipts(loanId);
+            expect(noteReceipt.token).to.eq(mockERC20.address);
+            expect((noteReceipt).amount).to.eq(ethers.utils.parseEther("110"));
+        });
+
+        it("Force repay full. 100 ETH principal, 10% interest rate, half duration.", async () => {
+            const { repaymentController, vaultFactory, mockERC20, loanCore, borrower, blockchainTime } = ctx;
+
+            const { loanId, bundleId } = await initializeLoan(
+                ctx,
+                mockERC20.address,
+                BigNumber.from(31536000), // durationSecs
+                ethers.utils.parseEther("100"), // principal
+                1000, // interest
+                1754884800, // deadline
+            );
+
+            // total repayment amount
+            const total = ethers.utils.parseEther("105");
+            const repayAdditionalAmount = total.sub(await mockERC20.balanceOf(borrower.address));
```

Comment:
For tests where we do not pre-calculate the interest amount, this pattern of minting the 'additionalAmount' works well when we need to check the balance of the borrower after the tx happens. Since the borrower receives the principal when the loan is started, minting the total logically doesnt make sense, where are minting the difference does so that our borrower balances are exact.

 It is also used in places where there are origination fees so you dont have to calculate the fees it will mint the difference to the borrower and the expect statements later are more straight forward.